### PR TITLE
Add ReStructuredText descriptor for console component.

### LINKED ISSUE #45717: Add RST descriptor for console component.
### Description

The console component currently has descriptors for Markdown, Json, etc. This feature would add RST as an option.

### Example

    $helper = new DescriptorHelper();
    $helper->describe($output, $this->getApplication(), [
      'format' => 'rst',
    ]);

### PR DESCRIPTION
**Pull request by grasmash**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #45717
| License       | MIT
| Doc PR        | n/a, descriptors are not documented.

### Description

The console component currently has descriptors for Markdown, Json, etc. This feature would add Restructured Text descriptor (rst) as an option.

### Example

    $helper = new DescriptorHelper();
    $helper->describe($output, $this->getApplication(), [
      'format' => 'rst',
    ]);

### CODE REVIEW

**Comment by grasmash** [comment]
Not sure why the Psalm test isn't passing. It doesn't seem to have anything to do with the changes in this PR. I find the output of that job hard to parse.
Similarly, the Appveyor failure is unrelated:
```
1) Symfony\Component\Messenger\Bridge\Redis\Tests\Transport\RedisExtIntegrationTest::testGetAfterReject
Failed asserting that null is not null.
C:\projects\symfony\src\Symfony\Component\Messenger\Bridge\Redis\Tests\Transport\RedisExtIntegrationTest.php:316
```

I think this is ready for review.

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+use Symfony\Component\Console\Application;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Helper\Helper;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputDefinition;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * ReStructuredTextDescriptor descriptor.
```
Comment (OskarStark) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php:23]: Can be removed

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        parent::write($content, $decorated);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function describeInputArgument(InputArgument $argument, array $options = [])
+    {
+        $this->write(
+      '``'.($argument->getName() ?: '<none>')."``\n".str_repeat($this->argumentTitleUnderlineChar, Helper::width($argument->getName()) + 4)."\n\n"
```
Comment (fabpot) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php:59]: indentation looks wrong

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        $name = '--'.$option->getName();
+        if ($option->isNegatable()) {
+            $name .= '|--no-'.$option->getName();
+        }
+        if ($option->getShortcut()) {
+            $name .= '|-'.str_replace('|', '|-', $option->getShortcut()).'';
+        }
+
+        $this->write(
+      '``'.$name.'``'."\n".str_repeat($this->argumentTitleUnderlineChar, Helper::width($name) + 4)."\n\n"
```
Comment (fabpot) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php:81]: here as well

**Review by fabpot — CHANGES_REQUESTED** [review]
LGTM

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+        $output->setDecorated($decorated);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function write(string $content, bool $decorated = true)
+    {
+        parent::write($content, $decorated);
```
Comment (stof) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php:50]: why overriding the method to do nothing else than calling the parent ?

File: src/Symfony/Component/Console/Tests/Fixtures/application_1.rst [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+
+Options
+=======
+
+``--format``
+------------
+
+The output format (txt, xml, json, or md)
+
+- Accept value: yes
```
Comment (stof) [inline src/Symfony/Component/Console/Tests/Fixtures/application_1.rst:177]: should be `Accepts value` to be consistent with the usage of `Is` below

File: src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+use Symfony\Component\Console\Input\InputDefinition;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * ReStructuredTextDescriptor descriptor.
+ */
+class ReStructuredTextDescriptor extends Descriptor
+{
+    private $applicationTitleUnderlineChar = '#';
```
Comment (javiereguiluz) [inline src/Symfony/Component/Console/Descriptor/ReStructuredTextDescriptor.php:27]: Some context in case anyone needs it: RST format doesn't define which characters to use for `<h1>`, `<h2>`, `<h3>`, etc. RST defines a series of characters that you can use (`###`, `===`, `^^^`, `~~~`, etc.) and the first one found is considered `<h1>`, the second set of characters found is considered `<h2>`, etc.

-----

I'm not saying that we need to do the same here, but, for your consideration, the standard notation that we follow in Symfony Docs is:

```
h1 =====
h2 -----
h3 ~~~~~
h4 .....
```

Here we would be using:

```
h1 #####
h2 *****
h3 =====
h4 -----
```

File: src/Symfony/Component/Console/Tests/Fixtures/application_1.rst [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+
+- ``completion [--debug] [--] [<shell>]``
+
+The completion command dumps the shell completion script required
+to use shell autocompletion (currently only bash completion is supported).
+
+Static installation
+-------------------
+
+Dump the script to a global completion file and restart your shell:
```
Comment (javiereguiluz) [inline src/Symfony/Component/Console/Tests/Fixtures/application_1.rst:24]: I haven't checked if this works too, but I think that "literal blocks" require:

* Contents to be indented by 4 spaces (this works perfectly in your code)
* The line before the literal block must end with two colons `::` (<-- this is the missing piece here)

See https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#literal-blocks

**Review by javiereguiluz — COMMENTED** [review]
Nice feature @grasmash! Thanks for contributing it.

File: src/Symfony/Component/Console/Tests/Fixtures/application_1.rst (hunk 7 above)
Comment (wouterj) [reply src/Symfony/Component/Console/Tests/Fixtures/application_1.rst:24]: Indeed, the line before the literal block *must* end with `::`. See also the official reStructuredText markup specification: https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#literal-blocks

**Comment by fabpot** [comment]
@grasmash Are you still interested in finishing this PR?

**Comment by fabpot** [comment]
Closing as stalled. Don't hesitate to reopen when you're ready.

**Comment by danepowell** [comment]
I've cleaned this up, addressed feedback, and re-opened as #48981
